sound/soc/sof/loader.c

```diff
@@ -129,6 +130,82 @@ int snd_sof_fw_parse_ext_data(struct snd_sof_dev *sdev, u32 bar, u32 offset)
 }
 EXPORT_SYMBOL(snd_sof_fw_parse_ext_data);
 
+static size_t snd_sof_ext_man_size(const struct firmware *fw)
+{
+	const struct ext_man_header *head = (void *)fw->data;
+
+	if (fw->size < sizeof(*head))
+		return 0;
+
+	/*
+	 * When fw points to extended manifest,
+	 * then first u32 must be equal EXT_MAN_MAGIC_NUMBER.
+	 */
+	if (head->magic == EXT_MAN_MAGIC_NUMBER)
+		return head->full_size;
+
+	return 0;
+}
+
+/* parse extended FW manifest data structures */
+static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+				    const struct firmware *fw)
+{
+	const struct ext_man_header *head = (struct ext_man_header *)fw->data;
+	const struct ext_man_elem_header *elem_hdr;
+	ssize_t remaining;
+	uintptr_t iptr;
+	int ret = 0;
+
+	remaining = head->full_size - head->header_size;
+
+	/* Assert firmware starts with extended manifest */
+	if (snd_sof_ext_man_size(fw) == 0)
+		return -EINVAL;
```

Comment:
well then return -EINVAL directly in ext_man_size() and check for ret < 0

zero meaning errors is opposite to the kernel conventions. Zero is success usually.

Reply:
> Zero is success usually.

But size 0 is size 0. Function name explicit describe what is returning, zero size is a case for FW without ext_man.

Reply:
the point is that you can return a value

< 0. that's an error case
== 0. that's just a case without extended manifest, not an error
>0. valid size w/ extended manifest.

It's better to keep errors and 'not errors' separate.

Reply:
changed